doc/port-fec/fec-auto-determination.md

```diff
@@ -0,0 +1,187 @@
+1. DPB use case: Enhance today's DPB CLI to automatically determine and configure FEC in CONFIG_DB for dynamically created ports, based on determine-fec module.
+2. non-DPB use case: Add a user-triggered CLI `fec-auto-correct` to automatically determine and configure FEC in CONFIG_DB for existing ports, based on determine-fec module.
+    - Future plan: determine-fec module can be further enhanced to be integrated with xcvrd, which can be triggered automatically during transceiver insertion, without human intervention. (details TBD)
+
+### API design
+```
+def determine_fec(lane_speed: int, num_lanes: int, optics_type: Optional[str] = None) -> str:
+    """
+    Determines the appropriate Forward Error Correction (FEC) type based on lane speed, number of lanes, and optics type for a specific logical port.
+    This logic is based on FEC mapping rules common for all platforms.
+
+    Parameters:
+    - lane_speed (int): The speed of each lane in GB.
+    - num_lanes (int): The total number of lanes.
+    - optics_type (Optional[str]): The type of optics in use. Can be None if not applicable.
+
+    Returns:
+    - str: The recommended FEC type based on the common rules. It can be either 'none'/'rs'/'fc'.
+
+    Example:
+    >>> determine_fec(25, 4, "100G-SR4")
+    "rs"
+
+    """
+```
+
+### Common Rule for FEC Determination
+
+#### Table 1: FEC Mapping Based on Optics Type
+| Optics Type | FEC  |
+|-------------|------|
+| 40G         | none |
+| 100G-DR     | none |
+| 100G-FR     | none |
+| 100G-LR     | none |
+| 100G-LR4    | none |
+| 100G-ER4    | none |
+| 100G AOC    | none |
+| 400G        | rs   |
+| ALL_OTHER   | rs   |
+
+#### Table 2: FEC Mapping Based on Lane Speed and Number of Lanes
+| Lane Speed | Number of Lanes (per Logical Port) | FEC  |
+|------------|-----------------------------------|------|
+| 10         | 1                                 | none |
+| 10         | 4                                 | none |
+| 20         | 2                                 | none |
+| 25         | 1                                 | rs   |
+| 25         | 2                                 | rs   |
+| 25         | 4                                 | rs   |
+| 25         | 8                                 | rs   |
+| 50         | 1                                 | rs   |
+| 50         | 2                                 | rs   |
+| 50         | 4                                 | rs   |
+| 50         | 8                                 | rs   |
+| 50         | 16                                | rs   |
+
+Above tables can be defined as JSON file, and be loaded by determine-fec module. Platform can also override the default FEC mapping by providing its own JSON file.
+
+> [!NOTE]
+> If a port has matched FEC entry in both above tables, then prefers FEC entry in first table. (Only exception: For `lane_speed=10, num_lane=1 or 4`, we prefer FEC entry in second table)
```

Comment:
First table defines the Optical media interface only. there can be on a single optical interface different types of electrical interface which define different FEC types.
for example: 100GBASE-DR (Clause 140) defines optical interface of 1 fiber lane.
Electrical interfaces that can be are (according to IEEE clause 140)
CAUI-4 (100G on 4 electrical lanes) - NO-FEC
100GAUI-2 (100G on 2 electrical lanes) - RS-FEC
100GAUI-1 (100G on 1 electrical lane) - RS FEC

This does not align with the statement in line 107 that " If a port has matched FEC entry in both above tables, then prefers FEC entry in first table".
FEC should be decided based on the Electrical interface (Second table) and not the first table in optical cable cases

Reply:
I agree, there are some cases where on same lane_speed you can have different FEC results.
so it should based on inputs from both tables, but not just one of them